### PR TITLE
implement out variant for softmax

### PR DESCRIPTION
**Pull request by makslevental**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61824
* __->__ #61823
* #61821
* #61802
* #61801
* #61800

Differential Revision: [D29812527](https://our.internmc.facebook.com/intern/diff/D29812527)

### CODE REVIEW

**Review by desertfire — APPROVED** [review]
LGTM

File: aten/src/ATen/native/SoftMax.cpp [hunk 1]
```diff
@@ -154,7 +147,14 @@ Tensor log_softmax_cpu(const Tensor& input_, const int64_t dim_, const bool half
       c10::nullopt /* layout */,
       c10::nullopt /* device */,
       c10::nullopt /* pin_memory */,
-      LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+      at::MemoryFormat::Contiguous);
```
Comment (desertfire) [inline aten/src/ATen/native/SoftMax.cpp:150]: Found this, https://fb.workplace.com/groups/pytorch.dev/permalink/543001792944972/. Not sure if revisiting LEGACY_CONTIGUOUS_MEMORY_FORMAT is still on the plan, but you may want to keep the format as is. @eellison @Krovatkin, do you guys know any history about this?

Comment (makslevental) [reply aten/src/ATen/native/SoftMax.cpp:150]: ah gotcha. i misunderstood that `define` i'll undo it.

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812527).

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812527).
